File: src/exception/timer/timer.c
#include "timer.h"
#include "gic/gic.h"
#include "sched/sched.h"
#include "uart/uart.h"

// All four are read/written from IRQ context as well as task context, so
// mark them volatile to keep the compiler honest.
static volatile uint64_t timer_freq = 0;
static volatile uint64_t timer_interval = 0;
static volatile uint64_t tick_count = 0;
static volatile timer_callback_t tick_callback = 0;

void timer_init() {
  uart_println("[TIMER] Initializing hardware timer");
  // initial value set by fw
  __asm__ __volatile__("mrs %0, cntfrq_el0" : "=r"(timer_freq));

  uart_printf("[TIMER] Frequency: %u Hz (%u MHz)\n", timer_freq,
              timer_freq / 1000000);

  gic_enable_irq(TIMER_PPI_INTID);

  uart_println("[TIMER] Initialized!");
}

void timer_start(uint64_t interval_ms) {
  if (timer_freq == 0) {
    uart_errorln("[TIMER] Not initialized! Call timer_init() first");
    return;
  }

  // milliseconds to timer ticks
  timer_interval = timer_freq * interval_ms / 1000;
  tick_count = 0;

  uart_printf("[TIMER] Starting with interval: %u ms (%u ticks)\n", interval_ms,
              timer_interval);

  // Use absolute deadlines (CVAL) instead of relative countdowns (TVAL)
  // so IRQ latency does not accumulate over time. CVAL is compared against
  // CNTPCT_EL0 by the timer; once they cross, the IRQ fires.
  uint64_t now;
  __asm__ __volatile__("mrs %0, cntpct_el0" : "=r"(now));
  __asm__ __volatile__(
      "msr cntp_cval_el0, %0" ::"r"(now + timer_interval));    // Deadline
  __asm__ __volatile__("msr cntp_ctl_el0, %0" ::"r"(1ULL));    // Enable

  uart_println("[TIMER] Started!");
}

void timer_stop() {
  __asm__ __volatile__("msr cntp_ctl_el0, %0" ::"r"(0ULL));
  uart_printf("[TIMER] Stopped after %u ticks\n", tick_count);
}

void timer_handle_irq() {
  tick_count++;

  // Re-arm by advancing the absolute deadline. This drifts at most one
  // interval per IRQ no matter how long handler latency takes; using
  // CNTP_TVAL_EL0 here would reset the countdown from "now" and accumulate.
  uint64_t cval;
  __asm__ __volatile__("mrs %0, cntp_cval_el0" : "=r"(cval));
  cval += timer_interval;
  __asm__ __volatile__("msr cntp_cval_el0, %0" ::"r"(cval));

  // Wake any tasks whose sleep deadline has passed
  sched_wake_sleepers();

  if (tick_callback) {
    tick_callback();
  } else {
    // Only log every 100 ticks (1 second at 10ms interval) to avoid spam
    if (tick_count % 100 == 0) {
      uart_printf("[TIMER] tick %u\n", tick_count);
    }
  }
}

void timer_set_callback(timer_callback_t cb) { tick_callback = cb; }

uint64_t timer_get_frequency() { return timer_freq; }

uint64_t timer_get_count() {
  uint64_t count;
  // Use the *physical* counter to match the physical timer (CNTPCT_EL0).
  // CNTVCT_EL0 (virtual) can diverge from CNTPCT_EL0 under virtualization.
  __asm__ __volatile__("mrs %0, cntpct_el0" : "=r"(count));
  return count;
}

uint64_t timer_get_ticks() { return tick_count; }

uint64_t timer_uptime_ms(void) { return tick_count * TIMER_INTERVAL_MS; }
uint64_t timer_uptime_seconds(void) {
  return (tick_count * TIMER_INTERVAL_MS) / 1000;
}

File: src/exception/timer.rs
//! ARM Generic Timer — periodic tick via the physical timer (PPI INTID 30).
//!
//! Uses absolute deadlines (CNTP_CVAL_EL0) re-armed each IRQ so latency does not
//! accumulate. The tick drives the scheduler: `handle_irq` wakes sleeping tasks
//! (via a registered hook) and invokes an optional tick callback.

use crate::exception::gic;
use crate::klib::sync::SpinLockIrqSafe;
use crate::kprintln;
use crate::{mrs, msr};

/// Private-peripheral interrupt ID for the EL1 physical timer on QEMU virt.
pub const TIMER_PPI_INTID: u32 = 30;
/// Default tick interval: 10 ms for fine-grained sleep.
pub const TIMER_INTERVAL_MS: u64 = 10;

struct Timer {
    freq: u64,
    interval: u64,
    tick_count: u64,
    callback: Option<fn()>,
    /// Hook the scheduler registers to wake sleeping tasks each tick.
    wake_sleepers: Option<fn()>,
}

// IRQ-safe: TIMER is locked both from task context (sleep_ms / uptime_ms via a
// syscall, IRQs unmasked) and from the timer IRQ handler. A plain SpinLock here
// deadlocks if a tick lands while a syscall holds the lock.
static TIMER: SpinLockIrqSafe<Timer> = SpinLockIrqSafe::new(Timer {
    freq: 0,
    interval: 0,
    tick_count: 0,
    callback: None,
    wake_sleepers: None,
});

/// Initialize the timer: read CNTFRQ and enable its PPI in the GIC.
pub fn init() {
    kprintln!("[TIMER] Initializing hardware timer");
    let freq = mrs!("cntfrq_el0");
    TIMER.lock().freq = freq;
    kprintln!(
        "[TIMER] Frequency: {} Hz ({} MHz)",
        freq,
        freq / 1_000_000
    );

    gic::enable_irq(TIMER_PPI_INTID);
    kprintln!("[TIMER] Initialized!");
}

/// Start periodic ticks at `interval_ms`.
pub fn start(interval_ms: u64) {
    let interval;
    {
        let mut t = TIMER.lock();
        if t.freq == 0 {
            crate::klib::uart::Uart.errorln("[TIMER] Not initialized! Call init() first");
            return;
        }
        t.interval = t.freq * interval_ms / 1000;
        t.tick_count = 0;
        interval = t.interval;
    }

    kprintln!(
        "[TIMER] Starting with interval: {} ms ({} ticks)",
        interval_ms,
        interval
    );

    // Absolute-deadline arming: CVAL = now + interval, then enable.
    let now = mrs!("cntpct_el0");
    unsafe {
        msr!("cntp_cval_el0", now + interval);
        msr!("cntp_ctl_el0", 1);
    }
    kprintln!("[TIMER] Started!");
}

/// Stop the timer.
pub fn stop() {
    unsafe {
        msr!("cntp_ctl_el0", 0);
    }
    let ticks = TIMER.lock().tick_count;
    kprintln!("[TIMER] Stopped after {} ticks", ticks);
}

/// Timer IRQ handler: advance the deadline, wake sleepers, run the callback.
pub fn handle_irq() {
    let (interval, callback, wake, count) = {
        let mut t = TIMER.lock();
        t.tick_count += 1;
        (t.interval, t.callback, t.wake_sleepers, t.tick_count)
    };

    // Re-arm by advancing the absolute deadline (bounded drift).
    let mut cval = mrs!("cntp_cval_el0");
    cval += interval;
    unsafe {
        msr!("cntp_cval_el0", cval);
    }

    if let Some(wake) = wake {
        wake();
    }

    if let Some(cb) = callback {
        cb();
    } else if count % 100 == 0 {
        // 1 s at a 10 ms interval; avoid per-tick spam.
        kprintln!("[TIMER] tick {}", count);
    }
}

pub fn set_callback(cb: fn()) {
    TIMER.lock().callback = Some(cb);
}

/// Register the scheduler's wake-sleepers hook (called once when sched starts).
pub fn set_wake_sleepers(hook: fn()) {
    TIMER.lock().wake_sleepers = Some(hook);
}

pub fn get_frequency() -> u64 {
    TIMER.lock().freq
}

/// Physical counter (CNTPCT_EL0).
pub fn get_count() -> u64 {
    mrs!("cntpct_el0")
}

pub fn get_ticks() -> u64 {
    TIMER.lock().tick_count
}

pub fn uptime_ms() -> u64 {
    TIMER.lock().tick_count * TIMER_INTERVAL_MS
}

pub fn uptime_seconds() -> u64 {
    (TIMER.lock().tick_count * TIMER_INTERVAL_MS) / 1000
}
